internal/services/mssql/mssql_server_resource.go

```diff
@@ -340,6 +341,9 @@ func resourceMsSqlServerUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
 				return fmt.Errorf("expanding `identity`: %+v", err)
 			}
 			payload.Identity = expanded
+		} else if strings.EqualFold(string(payload.Identity.Type), "SystemAssigned,UserAssigned") {
+			// switch out the legacy API value (no space) for the value used in the Schema (w/space for consistency)
+			payload.Identity.Type = identity.TypeSystemAssignedUserAssigned
```

Comment:
The Marshal function already handles this: https://github.com/hashicorp/go-azure-helpers/blob/5b8e01f22f0c7713bf4660645c102916bf71bb7c/resourcemanager/identity/legacy_system_and_user_assigned_map.go#L36-L39

Therefore the bug to fix here is going to be that we need to account for this when Unmarshaling the value in `hashicorp/go-azure-helpers` - rather than adding a resource-specific hack here?

Reply:
> Therefore the bug to fix here is going to be that we need to account for this when Unmarshaling the value in `hashicorp/go-azure-helpers` - rather than adding a resource-specific hack here?

Hi @tombuildsstuff thanks for your feedback. However, when I debugged, I found that the internal error occurred at https://github.com/hashicorp/go-azure-helpers/blob/5b8e01f22f0c7713bf4660645c102916bf71bb7c/resourcemanager/identity/legacy_system_and_user_assigned_map.go#L30. I don't quite understand what "we need to account for this when Unmarshaling the value in hashicorp/go-azure-helpers" means, could you please explain it?

Reply:
@sinbai

Sure. So from our side we use the consistent value `SystemAssigned, UserAssigned` internally - and transform to/from the API value (`SystemAssigned,UserAssigned`) as required when Marshalling and Unmarshalling the JSON payload (that is, Writing [Marshalling] and Reading [Unmarshalling]) respectively.

Go marshals data using a `Marshal` function - and so [this Marshal function](https://github.com/hashicorp/go-azure-helpers/blob/5b8e01f22f0c7713bf4660645c102916bf71bb7c/resourcemanager/identity/legacy_system_and_user_assigned_map.go#L23-L57) is saying that when the field `Type` within the struct `LegacySystemAndUserAssignedMap` contains the value `SystemAssigned, UserAssigned`, when writing the JSON we should be outputting `SystemAssigned,UserAssigned` instead.

However whilst we're doing that when Marshalling the data, we're not doing that when Unmarshalling the data - and as such whilst the API might be returning:

```json
{
  "identity": {
    "type": "SystemAssigned,UserAssigned"
  }
}
```

We're parsing this value out as `SystemAssigned,UserAssigned` rather than `SystemAssigned, UserAssigned` - as such we'll want to add a custom `Unmarshal` function to ensure we parse the correct value for the `Type` field.

Whilst [the Marshal function contains a check for `SystemAssigned,UserAssigned`](https://github.com/hashicorp/go-azure-helpers/blob/5b8e01f22f0c7713bf4660645c102916bf71bb7c/resourcemanager/identity/legacy_system_and_user_assigned_map.go#L29-L31), which should be being hit in this scenario, it's not being hit - and as such I suspect that's ultimately where this bug is manifesting - that we're not sending a value for the `type` field - hence the `400 Bad Request` being returned from the API. You should be able to see this through the API Requests/Responses in the Debug logs.

As such, to fix this, we'll want to implement a custom Unmarshal function, similar to below - in `hashicorp/go-azure-helpers` to test this:

```go
func (s *LegacySystemAndUserAssignedMap) UnmarshalJSON(input []byte) error {
	if input == nil {
		return nil
	}

	var temp map[string]interface{}
	if err := json.Unmarshal(input, &temp); err != nil {
		return fmt.Errorf("unmarshaling LegacySystemAndUserAssignedMap into map[string]interface: %+v", err)
	}
	typeVal := TypeNone
	if v, ok := temp["type"].(string); ok && v != "" {
		if strings.EqualFold(v, string(TypeSystemAssigned)) {
			typeVal = TypeSystemAssigned
		}
		if strings.EqualFold(v, string(TypeUserAssigned)) {
			typeVal = TypeUserAssigned
		}

		if strings.EqualFold(v, string(typeLegacySystemAssignedUserAssigned)) {
			typeVal = TypeSystemAssignedUserAssigned
		}
		if strings.EqualFold(v, string(TypeSystemAssignedUserAssigned)) {
			typeVal = TypeSystemAssignedUserAssigned
		}
	}

	type alias LegacySystemAndUserAssignedMap
	var decoded alias
	if err := json.Unmarshal(input, &decoded); err != nil {
		return fmt.Errorf("unmarshaling: %+v", err)
	}

	s.Type = typeVal
	s.IdentityIds = decoded.IdentityIds
	s.PrincipalId = decoded.PrincipalId
	s.TenantId = decoded.TenantId

	return nil
}
```

Reply:
@tombuildsstuff thank you very much for the clear explanation. I will close this PR and fix it in go-azure-helpers as soon as possible.